### PR TITLE
Fix query serialization round-trip issue

### LINKED ISSUE #2585: SPARQL query round-trip serialization error
### Version

5.0.0

### What happened?

We're building a SPARQL query programatically and then serializing the query to a string to be able to run it against a Fuseki SPARQL endpoint. In one case, we build a query with a valid URI node that is then serialized as an invalid prefixed name, because the local part starts with an invalid character according to https://www.w3.org/TR/sparql11-query/#rPN_LOCAL

This query exhibits the issue:

```sparql
PREFIX eg: <http://example.com/with-trailing/>

ASK { <http://example.com/with-trailing/-with_leading-> a eg:BadPN_LOCAL . }
```

Using `qparse` to parse the above throws a `QueryParseException` as below when it checks the output query by re-parsing it.

### Relevant output and stacktrace

```shell
$ qparse 'PREFIX eg: <http://example.com/with-trailing/> ASK { <http://example.com/with-trailing/-with_leading-> a eg:BadPN_LOCAL . }'
PREFIX  eg:   <http://example.com/with-trailing/>

ASK
WHERE
  { eg:-with_leading-
              a  eg:BadPN_LOCAL
  }


**** Check failure: could not parse output query
org.apache.jena.query.QueryParseException: Encountered " "-" "- "" at line 5, column 8.
Was expecting one of:
    <IRIref> ...
    <PNAME_NS> ...
    <PNAME_LN> ...
    <VAR1> ...
    <VAR2> ...
    "a" ...
    "distinct" ...
    "multi" ...
    "shortest" ...
    "(" ...
    "!" ...
    "^" ...
    
        at org.apache.jena.sparql.lang.ParserARQ.perform(ParserARQ.java:99)
        at org.apache.jena.sparql.lang.ParserARQ.parse$(ParserARQ.java:52)
        at org.apache.jena.sparql.lang.SPARQLParser.parse(SPARQLParser.java:33)
        at org.apache.jena.query.QueryFactory.parse(QueryFactory.java:144)
        at org.apache.jena.query.QueryFactory.create(QueryFactory.java:83)
        at org.apache.jena.sparql.util.QueryUtils.checkParse(QueryUtils.java:96)
        at org.apache.jena.sparql.util.QueryUtils.checkQuery(QueryUtils.java:38)
        at arq.qparse.exec(qparse.java:158)
        at org.apache.jena.cmd.CmdMain.mainMethod(CmdMain.java:87)
        at org.apache.jena.cmd.CmdMain.mainRun(CmdMain.java:56)
        at org.apache.jena.cmd.CmdMain.mainRun(CmdMain.java:43)
        at arq.qparse.main(qparse.java:66)
PREFIX  eg:   <http://example.com/with-trailing/>

ASK
WHERE
  { eg:-with_leading-
              a  eg:BadPN_LOCAL
  }
```

### Are you interested in making a pull request?

Yes

### PR DESCRIPTION
**Pull request by ajtucker**
GitHub issue resolved #2585

Pull request Description:

Re-do char-by-char local PName validation following grammar rules at https://www.w3.org/TR/sparql11-query/#rPN_LOCAL

----

 - [X] Tests are included.
 - [ ] Documentation change and updates are provided for the [Apache Jena website](https://github.com/apache/jena-site/)
 - [ ] Commits have been squashed to remove intermediate development commit messages.
 - [ ] Key commit messages start with the issue number (GH-xxxx)

By submitting this pull request, I acknowledge that I am making a contribution to the Apache Software Foundation under the terms and conditions of the [Contributor's Agreement](https://www.apache.org/licenses/contributor-agreements.html).

----

See the [Apache Jena "Contributing" guide](https://github.com/apache/jena/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: jena-arq/src/main/java/org/apache/jena/sparql/util/FmtUtils.java [hunk 1]
```diff
@@ -477,27 +477,56 @@ private static boolean checkValidLocalname(String localname) {
...
+        return validPNameChar(ch) || ( ch == '.' ) || ( ch == ':' ) || ( ch == '%' ) ;
+    }
 
-        return true;
+    // last char, but not first, in last production in 169: https://www.w3.org/TR/sparql11-query/#rPN_LOCAL
+    private static boolean validEndPNameChar(char ch) {
+        return validPNameChar(ch) || ( ch == ':' ) || ( ch == '%' ) ;
+    }
+
+    // production 164: https://www.w3.org/TR/sparql11-query/#rPN_CHARS_BASE
```
Comment (afs) [inline jena-arq/src/main/java/org/apache/jena/sparql/util/FmtUtils.java:509]: There are functions to test for the token rules in `RiotChars`.

This one is `RiotChars.isPNCharsBase`.

File: jena-arq/src/main/java/org/apache/jena/sparql/util/FmtUtils.java [hunk 2]
```diff
@@ -477,27 +477,56 @@ private static boolean checkValidLocalname(String localname) {
...
 
+    // production 167: https://www.w3.org/TR/sparql11-query/#rPN_CHARS
     private static boolean validPNameChar(char ch) {
-        if ( Character.isLetterOrDigit(ch) ) return true ;
-        if ( ch == '.' )    return true ;
-        if ( ch == ':' )    return true ;
-        if ( ch == '-' )    return true ;
-        if ( ch == '_' )    return true ;
-        return false ;
+        return validPNameCharU(ch) ||
```
Comment (afs) [inline jena-arq/src/main/java/org/apache/jena/sparql/util/FmtUtils.java:527]: This is `RiotChars.isPNChars`.

File: jena-arq/src/main/java/org/apache/jena/sparql/util/FmtUtils.java [hunk 3]
```diff
@@ -477,27 +477,56 @@ private static boolean checkValidLocalname(String localname) {
...
+                (ch >= '\u0370' && ch <= '\u037D') || (ch >= '\u037F' && ch <= '\u1FFF') ||
+                (ch >= '\u200C' && ch <= '\u200D') || (ch >= '\u2070' && ch <= '\u218F') ||
+                (ch >= '\u2C00' && ch <= '\u2FEF') || (ch >= '\u3001' && ch <= '\uD7FF') ||
+                (ch >= '\uF900' && ch <= '\uFDCF') || (ch >= '\uFDF0' && ch <= '\uFFFD');
+        // 32bit unicode chars are 2 Java chars ( ch >= '\u10000' && ch <= '\uEFFFF' )
+    }
+
+    // production 165: https://www.w3.org/TR/sparql11-query/#rPN_CHARS_U
+    private static boolean validPNameCharU(char ch) {
+        return validPNameCharBase(ch) || ( ch == '_' ) ;
```
Comment (afs) [inline jena-arq/src/main/java/org/apache/jena/sparql/util/FmtUtils.java:522]: This is `RiotChars.isPNChars_U`.

**Review by afs — CHANGES_REQUESTED** [review]
Sorry for the delay in getting to this PR.

There functions for the tests in `RiotChars`. Having one place is better and also the `RiotChars` are not dependent of what Java defines as a letter or digit which is Unicode version dependent.

Other than that, the PR looks good to me. Thank you.

**Comment by ajtucker** [comment]
> There functions for the tests in `RiotChars`. Having one place is better and also the `RiotChars` are not dependent of what Java defines as a letter or digit which is Unicode version dependent.

Thanks. I missed these when looking, being thrown by suspecting that the parser source had been generated.

**Review by afs — APPROVED** [review]

**Pull request merged**